stablehlo/reference/Element.cpp

```diff
@@ -0,0 +1,164 @@
+  APFloat err(f.getSemantics(), "1e-5");
+  return absFloat.compare(err) == APFloat::cmpLessThan;
+}
+
+bool isSupportedUnsignedIntegerType(Type type) {
+  return type.isUnsignedInteger(4) || type.isUnsignedInteger(8) ||
+         type.isUnsignedInteger(16) || type.isUnsignedInteger(32) ||
+         type.isUnsignedInteger(64);
+}
+
+bool isSupportedSignedIntegerType(Type type) {
+  // TODO(#22): StableHLO, as bootstrapped from MHLO, inherits signless
+  // integers which was added in MHLO for legacy reasons. Going forward,
+  // StableHLO will adopt signfull integer semantics with signed and unsigned
+  // integer variants.
+  return type.isSignlessInteger(4) || type.isSignlessInteger(8) ||
+         type.isSignlessInteger(16) || type.isSignlessInteger(32) ||
+         type.isSignlessInteger(64);
+}
+
+bool isSupportedIntegerType(Type type) {
+  return isSupportedUnsignedIntegerType(type) ||
+         isSupportedSignedIntegerType(type);
+}
+
+bool isSupportedFloatType(Type type) {
+  return type.isF16() || type.isBF16() || type.isF32() || type.isF64();
+}
+
+bool isSupportedComplexType(Type type) {
+  auto complexTy = type.dyn_cast<mlir::ComplexType>();
+  if (!complexTy) return false;
+
+  auto complexElemTy = complexTy.getElementType();
+  return complexElemTy.isF32() || complexElemTy.isF64();
+}
+
+APFloat getFloatValue(Element element) {
+  return element.getValue().cast<FloatAttr>().getValue();
+}
+
+APInt getIntegerValue(Element element) {
+  return element.getValue().cast<IntegerAttr>().getValue();
+}
+
+std::complex<APFloat> getComplexValue(Element element) {
+  auto arryOfAttr = element.getValue().cast<ArrayAttr>().getValue();
+  return std::complex<APFloat>(arryOfAttr[0].cast<FloatAttr>().getValue(),
+                               arryOfAttr[1].cast<FloatAttr>().getValue());
+}
+
+}  // namespace
+
+Element Element::operator+(const Element &other) const {
+  auto type = getType();
+  assert(type == other.getType());
+
+  if (isSupportedFloatType(type)) {
+    auto left = getFloatValue(*this);
+    auto right = getFloatValue(other);
+    return Element(type, FloatAttr::get(type, left + right));
+  }
+
+  if (isSupportedIntegerType(type)) {
+    auto left = getIntegerValue(*this);
+    auto right = getIntegerValue(other);
+    return Element(type, IntegerAttr::get(type, left + right));
+  }
+
+  if (isSupportedComplexType(type)) {
+    auto complexElemTy = type.cast<ComplexType>().getElementType();
+    auto leftComplexValue = getComplexValue(*this);
+    auto rightComplexValue = getComplexValue(other);
+
+    return Element(
+        type,
+        ArrayAttr::get(
+            type.getContext(),
+            {FloatAttr::get(complexElemTy,
+                            leftComplexValue.real() + rightComplexValue.real()),
+             FloatAttr::get(complexElemTy, leftComplexValue.imag() +
+                                               rightComplexValue.imag())}));
+  }
+
+  llvm_unreachable("Unsupported element type");
```

Comment:
HI @joker-eph 

Our plan  was to have such  `llvm_unreachable` programmatic error  part of all the future operations over `Element` objects. IMO, having `llvm_unreachable`  simplifies the control flow as the error does not needs to be propagated up the stack.  With Expected<Element>,  we will loose the  following better readability while operating on the `Element` objects

Comparing

```
Tensor eval(AddOp op, const Tensor &lhs, const Tensor &rhs) {
  ...
  for (...) {
    **result.set(i, lhs.get(i) + rhs.get(i));**
  }
  return result;
}
```

Vs

```
Expected<Tensor> eval(AddOp op, const Tensor &lhs, const Tensor &rhs) {
  ...
  for (...) {
    if(auto lhsOrErr = lhs.get(i)) return ...some Error code .. 
    if(auto rhsOrErr = rhs.get(i)) return ...some Error code .. 
    result.set(i, *lhsOrErr + *rhsOrErr);
}
```

Let me know your opinion otherwise.

Reply:
I meant that instead of llvm_unreachable we could use llvm::report_fatal_error ; the only difference is that the later is effective without assertions enabled.

Reply:
Thanks for suggested! Addressed it in the current file and also in files `Tensor.cpp` and `Interpreter.cpp`